AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AWSAuthFederateToIdentityPoolOperation.swift

```diff
@@ -43,5 +45,109 @@ public class AWSAuthFederateToIdentityPoolOperation: AmplifyOperation<
             return
         }
 
+        authStateMachine.getCurrentState { [weak self] state in
+
+            guard case .configured(let authNState, let authZState) = state  else {
+                self?.sendInvalidStateError()
+                return
+            }
+
+            if self?.isValidAuthNStateToStart(authNState) == true &&
+               self?.isValidAuthZStateToStart(authZState) == true {
+                self?.startFederatingToIdentityPool()
+            } else {
+                self?.sendInvalidStateError()
+            }
+        }
+
+    }
+
+    func startFederatingToIdentityPool() {
+        var token: AuthStateMachine.StateChangeListenerToken?
+        token = authStateMachine.listen { [weak self] state in
+
+            guard  case .configured(let authNState, let authZState) = state else {
+                return
+            }
+
+            switch (authNState, authZState) {
+            case (.federatedToIdentityPool, .sessionEstablished(let credentials)):
+
+                self?.dispatch(credentials)
+                if let token = token {
+                    self?.authStateMachine.cancel(listenerToken: token)
+                }
+                self?.finish()
+            case (.error(_), .error(let authZError)):
+                self?.dispatch(AuthError.service(
+                    "Error federating to identity pool",
+                    AmplifyErrorMessages.shouldNotHappenReportBugToAWS(),
+                    authZError))
+                if let token = token {
+                    self?.authStateMachine.cancel(listenerToken: token)
+                }
+                self?.finish()
+            default:
+                break
+            }
+
+        } onSubscribe: { [weak self] in
+            self?.sendStartFederatingToIdentityPoolEvent()
+        }
+
+    }
+
+    func sendStartFederatingToIdentityPoolEvent() {
+        let federatedToken = FederatedToken(
+            token: request.token,
+            provider: request.provider)
+        let event = AuthorizationEvent.init(
+            eventType: .startFederationToIdentityPool(federatedToken))
+        authStateMachine.send(event)
+    }
+
+    func isValidAuthNStateToStart(_ authNState: AuthenticationState) -> Bool {
+        switch authNState {
+        case .configured, .notConfigured, .signedOut:
+            return true
+        default:
+            return false
+        }
+    }
+
+    func isValidAuthZStateToStart(_ authZState: AuthorizationState) -> Bool {
+        switch authZState {
+        case .configured, .sessionEstablished:
+            return true
+        default:
+            return false
+        }
+    }
+
+    func sendInvalidStateError() {
+        dispatch(AuthError.invalidState(
+            "No federation found.",
```

Comment:
We need to change this error message. Something along the lines: "Federation could not be completed, plugin is in an invalid state to continue federation" ?